scalameta/common/src/main/scala/org/scalameta/internal/logger.scala

```diff
@@ -0,0 +1,95 @@
+// The macro implementations are copied from https://github.com/lihaoyi/sourcecode
+package org.scalameta.internal
+
+import scala.language.existentials
+import scala.language.experimental.macros
+import scala.language.implicitConversions
+
+import scala.reflect.macros.blackbox
+
+abstract class SourceValue[T] {
+  def value: T
+}
+abstract class SourceCompanion[T, V <: SourceValue[T]](build: T => V) {
+  def apply()(implicit s: V): T = s.value
+  implicit def wrap(s: T): V    = build(s)
+}
+
+case class File(value: String) extends SourceValue[String]
+object File extends SourceCompanion[String, File](new File(_)) {
+  implicit def generate: File = macro impl
+
+  def impl(c: blackbox.Context): c.Expr[File] = {
+    import c.universe._
+    val file = c.enclosingPosition.source.path
+    c.Expr[File](q"""${c.prefix}($file)""")
+  }
+}
+
+case class Line(value: Int) extends SourceValue[Int]
+object Line extends SourceCompanion[Int, Line](new Line(_)) {
+  implicit def generate: Line = macro impl
+  def impl(c: blackbox.Context): c.Expr[Line] = {
+    import c.universe._
+    val line = c.enclosingPosition.line
+    c.Expr[Line](q"""${c.prefix}($line)""")
+  }
+}
+
+case class FileLine(file: File, line: Line)
+object FileLine {
+  implicit def generate(implicit file: File, line: Line): FileLine = FileLine(file, line)
+}
+
+case class Text[T](value: T, source: String)
+object Text {
+  implicit def generate[T](v: T): Text[T] = macro Impls.text[T]
+  def apply[T](v: T): Text[T] = macro Impls.text[T]
+}
+
+object Impls {
+  def text[T: c.WeakTypeTag](c: blackbox.Context)(v: c.Expr[T]): c.Expr[Text[T]] = {
+    import c.universe._
+    val fileContent = new String(v.tree.pos.source.content)
+    val start = v.tree.collect {
+      case treeVal =>
+        treeVal.pos match {
+          case NoPosition => Int.MaxValue
+          case p          => p.start
+        }
+    }.min
+    val g      = c.asInstanceOf[reflect.macros.runtime.Context].global
+    val parser = g.newUnitParser(fileContent.drop(start))
+    parser.expr()
+    val end  = parser.in.lastOffset
+    val txt  = fileContent.slice(start, start + end)
+    val tree = q"""${c.prefix}(${v.tree}, $txt)"""
+    c.Expr[Text[T]](tree)
+  }
+}
+
+object logger {
```

Comment:
How about putting this object in here? https://github.com/scalameta/scalameta/blob/e5136bed4289e98c2e852162d73f63a7b6988479/scalameta/common/src/main/scala/org/scalameta/package.scala#L29